### PR TITLE
Only store negotiated (not offered) SNI hostname in SSL_SESSION

### PR DESCRIPTION
**Pull request by kaduk**
Fix a thread-safety issue from #4519 by only writing to the SSL_SESSION during non-resumption handshakes and only when SNI is negotiated.

const-ify a few interfaces that can trivially do so (in order to use one of them), and also leave a comment noting work that can only be done at an API-breaking release boundary.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/s3_lib.c [hunk 1]
```diff
@@ -3466,6 +3466,25 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
...
+         * block of code cannot be enabled until ABI-breaking changes
+         * are permitted by release policy.
+         */
+        /*
+         * This is the API to set what SNI a client requests; it makes no
+         * sense for a server.
+         */
+        if (s->server) {
+            SSLerr(SSL_F_SSL3_CTRL, SSL_R_SSL3_EXT_INVALID_SERVERNAME);
+            return 0;
```
Comment (tmshort) [inline ssl/s3_lib.c]: Perhaps return success for this condition, but do nothing? The unit tests will then pass (or change the unit tests, but then this breaks existing behavior...)

Comment (davidben) [reply ssl/s3_lib.c]: The OpenSSL API is such that `s->server` is not meaningful in config functions. So you, in general, need to make your config functions always stash the value, but just ignore it if not applicable. E.g.

```
  SSL_CTX *ctx = SSL_CTX_new(TLS_method());
  SSL *ssl = SSL_new(ctx);
  SSL_set_something(ssl);
  SSL_set_something_else(ssl);
  // ssl->server is not set until this point:
  SSL_connect(ssl);  // Or SSL_accept(ssl)
```

Comment (kaduk) [reply ssl/s3_lib.c]: > The OpenSSL API is such that ```s->server``` is not meaningful in config functions.

The current one is, yes.  Perhaps it need not be, though (when we are permitted to make such breaking changes).  But I don't think there would be too much harm from just dropping the last commit, if I'm in the minority.  As a very far aside note, I've been pondering whether there is room to do a full-on API modernization for 1.2, making a new TLS structure and TLS_foo() public APIs that are free of historical cruft (such as this lack of meaning for ```s->server``` until the handshake begins).

I could also do as you implicitly suggest, and defer this check until SSL_do_handshake(), if people want that, of course.

Comment (davidben) [reply ssl/s3_lib.c]: Bifurcating all the APIs is kind of expensive, as folks tend to support a couple of OpenSSL versions at the time. At the least, before doing that, *definitely* provide a compatibility story. 1.1.0 is not a good model.

But you could fix this more simply. Just get rid of `TLS_method` and say that you want people to use `TLS_client_method` and `TLS_server_method`. I considered that for BoringSSL, but found that too much existing code used what was then `SSLv23_method` to desupport the uncommitted state. But with a proper deprecation path you could do it.

Though having the server setters "work" for clients and vice versa isn't the end of the world... except where OpenSSL has those problematic combined config and state getters. You've identified one of them in this PR. Those have time and time again demonstrated themselves to be an hindrance for improvements on our end. (We organize state not by whether it happens to be an extension, but by scope, which lets us easily, e.g., release all temporary handshake state after the handshake completes. Or cleanly support `SSL_clear`. But by clearly separating that state, it becomes very obvious each time some corresponding getter is ambiguous on what it actually does.)

File: ssl/statem/extensions.c [hunk 2]
```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
...
+     * storage in the SSL to the persistent SSL_SESSION, now that we
+     * know we accepted it.
+     * Clients make this copy when parsing the server's response to
+     * the extension, which is when they find out that the negotiation
+     * was successful.
+     */
+    if (s->server) {
+        if (!sent) {
+            OPENSSL_free(s->ext.hostname);
+            s->ext.hostname = NULL;
```
Comment (tdvi1) [inline ssl/statem/extensions.c]: It might be better to set s->ext.hostname to NULL at init_server_name()

Comment (tdvi1) [reply ssl/statem/extensions.c]: Ben, Can you please describe a scenario where sent is false, and s->ext.hostname isn't NULL ?
I can think of a scenario that involves the sni callback that sets s->ext.hostname,
but in such case, I don't think that s->ext.hostname should be set to NULL.

Comment (kaduk) [reply ssl/statem/extensions.c]: I believe in a TLS 1.2 resumption case where the client sends SNI and we successfully resume, the server will not have sent SNI in the ServerHello (```sent == 0```), but ```s->ext.hostname``` would have been set in ```tls_parse_ctos_server_name()```, so this free+NULL would have significance.

**Comment by kaduk** [comment]
@mattcaswell did you want to take a look at this?  Extra eyes on the ```SSL_get_servername()``` change in particular are probably warranted.

File: ssl/statem/extensions.c (hunk 2 above)
Comment (tdvi1) [reply ssl/statem/extensions.c]: 'sent' arg is 'raw_extension_st::present',
meaning that when final_server_name() is called at the server side (s->server is true),
then 'sent' doesn't mean that the server sent this extension.
'sent' means that the client sent this extension, and that the server received this extension.
so if sent is false, then s->ext.hostname is NULL, unless server name callback, or another callback has set it.
so it seems that this null assignment isn't needed.​

File: ssl/statem/extensions.c [hunk 3]
```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
...
-        s->session->ext.hostname = NULL;
+    /*
+     * For servers, propagate the SNI hostname from the temporary
+     * storage in the SSL to the persistent SSL_SESSION, now that we
+     * know we accepted it.
+     * Clients make this copy when parsing the server's response to
+     * the extension, which is when they find out that the negotiation
+     * was successful.
+     */
+    if (s->server) {
```
Comment (tdvi1) [inline ssl/statem/extensions.c]: When final_server_name() is called at the client side,
Should we consider coping the servername from s->ext.hostname to s->session->ext.hostname  ?
We may consider doing it, in case 'sent' is true. (At the client side, 'sent' is true when the client gets the "sni ack" from the server).

File: ssl/statem/extensions.c [hunk 4]
```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
...
+        if (!sent) {
+            OPENSSL_free(s->ext.hostname);
+            s->ext.hostname = NULL;
+        } else if (!s->hit || SSL_IS_TLS13(s)) {
+            /* Only store the hostname in the session if we accepted it. */
+            OPENSSL_free(s->session->ext.hostname);
+            s->session->ext.hostname = OPENSSL_strdup(s->ext.hostname);
```
Comment (tdvi1) [inline ssl/statem/extensions.c]: In think that in case of TLS 1.3 resumption,
when final_server_name() is called at the server side, then s->ext.hostname is currently NULL.
but s->session->ext.hostname may hold a value.
That is why we should consider making s->ext.hostname to hold the hostname at tls_parse_ctos_server_name,
even if s->hit is set.

File: ssl/statem/extensions.c (hunk 2 above)
Comment (kaduk) [reply ssl/statem/extensions.c]: This analysis looks reasonable within the bounds of a single handshake, but the situation gets more complicated when (pre-TLS 1.3) renegotiation is in play.  A hostname value could have been set by an initial handshake and then omitted from a renegotiation handshake, in which case this clearing code seems like it would still have non-trivial effect.

**Comment by kaduk** [comment]
@openssl/omc do you want to look at this now or do we think we need to write some tests that exercise the various renegotiation/resumption flows that have been discussed in the comments?

File: ssl/statem/extensions_srvr.c [hunk 5]
```diff
@@ -127,7 +127,7 @@ int tls_parse_ctos_server_name(SSL *s, PACKET *pkt, unsigned int context,
         return 0;
     }
 
-    if (!s->hit) {
+    if (!s->hit || SSL_IS_TLS13(s)) {
```
Comment (mattcaswell) [inline ssl/statem/extensions_srvr.c]: Why are we doing this if we have TLSv1.3?

File: ssl/statem/extensions_srvr.c [hunk 6]
```diff
@@ -142,21 +142,24 @@ int tls_parse_ctos_server_name(SSL *s, PACKET *pkt, unsigned int context,
             return 0;
         }
 
-        OPENSSL_free(s->session->ext.hostname);
-        s->session->ext.hostname = NULL;
-        if (!PACKET_strndup(&hostname, &s->session->ext.hostname)) {
+        /* Store the requested SNI in the SSL as temporary storage.
+         * If we accept it, it will get stored in the SSL_SESSION as well. */
```
Comment (mattcaswell) [inline ssl/statem/extensions_srvr.c]: Multi-line comment style.

File: ssl/s3_lib.c (hunk 1 above)
Comment (mattcaswell) [reply ssl/s3_lib.c]: I'm not sure I like the idea of adding `if 0` code here. It tends to rot. Just a simple comment would probably be better.

File: ssl/ssl_sess.c [hunk 7]
```diff
@@ -430,15 +430,6 @@ int ssl_get_new_session(SSL *s, int session)
...
 
-        if (s->ext.hostname) {
-            ss->ext.hostname = OPENSSL_strdup(s->ext.hostname);
-            if (ss->ext.hostname == NULL) {
-                SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL_GET_NEW_SESSION,
-                         ERR_R_INTERNAL_ERROR);
-                SSL_SESSION_free(ss);
-                return 0;
-            }
-        }
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: Why is this deleted?

File: ssl/s3_lib.c (hunk 1 above)
Comment (kaduk) [reply ssl/s3_lib.c]: To be honest, I wasn't even sure if any of the last commit would stick around.  But I can change it to just a comment.

File: ssl/ssl_sess.c (hunk 7 above)
Comment (kaduk) [reply ssl/ssl_sess.c]: This is in ssl_get_new_session(), where we don't want to just mechanically copy over the value from the SSL to the SSL_SESSION -- in the new information model, that only happens when the SNI value is actually negotiated for the connection.  (I only noticed it was here because some tests weren't behaving as I expected, which shows the value of tests, I suppose.)

File: ssl/statem/extensions_srvr.c (hunk 5 above)
Comment (kaduk) [reply ssl/statem/extensions_srvr.c]: TLS 1.3 changes (well, or not, depending on who you ask) the semantics of SNI from being bound to the session to being bound to the connection.  Semi-relatedly, we generate new SSL_SESSION objects for TLS 1.3 connections due to some other place we had to modify the structure (I forget the details offhand), so it's always safe to store the new SNI value in the per-connection SSL_SESSION object (which we do in final_server_name()).

File: ssl/statem/extensions_srvr.c (hunk 6 above)
Comment (kaduk) [reply ssl/statem/extensions_srvr.c]: ack

File: ssl/statem/extensions_srvr.c (hunk 5 above)
Comment (mattcaswell) [reply ssl/statem/extensions_srvr.c]: But how does this concretely change the behaviour? I can't work out how a server will behave different in TLSv1.3 vs TLSv1.2 after this change?

Comment (kaduk) [reply ssl/statem/extensions_srvr.c]: For TLS 1.3 resumptions where the client offers a new SNI, SSL_get_servername() will report the correct (new, per-connection) SNI value for TLS 1.3 connections.  If we did not have this chunk, that function would incorrectly report the old SNI value from the previous session/connection.  (Right?)

Comment (mattcaswell) [reply ssl/statem/extensions_srvr.c]: Ah. Right - thanks for the clarification. Where is this change of behaviour between TLSv1.2/TLSv1.3 documeted in the spec?

Comment (kaduk) [reply ssl/statem/extensions_srvr.c]: Section 4.2.11 "Pre-Shared Key Extension" has:

```
   In TLS versions prior to TLS 1.3, the Server Name Identification
   (SNI) value was intended to be associated with the session (Section 3
   of [RFC6066]), with the server being required to enforce that the SNI
   value associated with the session matches the one specified in the
   resumption handshake.  However, in reality the implementations were
   not consistent on which of two supplied SNI values they would use,
   leading to the consistency requirement being de facto enforced by the
   clients.  In TLS 1.3, the SNI value is always explicitly specified in
   the resumption handshake, and there is no need for the server to
   associate an SNI value with the ticket.  Clients, however, SHOULD
   store the SNI with the PSK to fulfill the requirements of
   Section 4.6.1.
```

There's also some text in NewSessionTicket about how resuming with a different SNI is a little dicey for now, but there's a draft in the works to say more about how to do it.

**Comment by mattcaswell** [comment]
This looks ok to me, but there are a couple of updates needed before approval: comment style and the "if 0" code.

**Comment by kaduk** [comment]
Thanks for the review!  I pushed this to master, so closing.

**Comment by dot-asm** [comment]
Extended test failure was apparently triggered by this, i.e. red cross at https://travis-ci.org/openssl/openssl/builds/406252276 appears relevant...
